### PR TITLE
CIRC-1548 NPE when creating TLR Recall for instance without an open loan

### PR DESCRIPTION
**Pull request by bisecomsergiy**
### Purpose
Fix occurring NPE when creating TLR Recall for instance without an open loan
### Approach
-	Update logic on tlr recall requests queues;
-	As per the ticket comments, implement logic for tlr recall requests for items with recallable statuses;
-	Add tests for recalls creation against items with recallable statuses
-	Add test to follow tickets requirements for hold requests creation

### Learning
https://issues.folio.org/browse/CIRC-1548

### CODE REVIEW

File: src/main/java/org/folio/circulation/resources/RequestFromRepresentationService.java [hunk 1]
```diff
@@ -275,11 +276,53 @@ private CompletableFuture<Result<Request>> fetchItemAndLoanForRecallTlrRequest(R
...
+      return failedValidation("Request doesn't have loan and items with allowed statuses",
+        "loan", null);
+    }
+    return of(() -> request);
+  }
+
+  private Item shouldHaveItemsWithAllowedStatusesForRecall(Request request) {
+    return request.getInstanceItems().stream()
+      .filter(item -> List.of("Awaiting delivery", "Paged", "Awaiting pickup")
+        .contains(item.getStatus().getValue()))
```
Comment (Mykyta-Varenyk) [inline src/main/java/org/folio/circulation/resources/RequestFromRepresentationService.java:323]: Extract the list into a variable please - the object is being created for every stream element

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 2]
```diff
@@ -1625,6 +1627,104 @@ private void recallItem(ItemResource item, List<UserResource> users) {
...
 
+  @Test
+  void tlrRecallShouldFailIfLoanAndAllowedStatusItemsForInstanceNotExist() {
+    configurationsFixture.enableTlrFeature();
+    var items = itemsFixture.createMultipleItemsForTheSameInstance(3);
+    var firstItem = items.get(0);
+    Response response = requestsFixture.attemptPlaceHoldOrRecallTLR(firstItem.getInstanceId(),
+      usersFixture.charlotte(), RECALL);
+
+    assertThat(response.getStatusCode(), CoreMatchers.is(422));
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/requests/RequestsAPICreationTests.java:1638]: static import
```suggestion
    assertThat(response.getStatusCode(), is(422));
```

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 3]
```diff
@@ -1625,6 +1627,104 @@ private void recallItem(ItemResource item, List<UserResource> users) {
       .forEach(index -> requestsFixture.recallItem(item, users.get(index)));
   }
 
+  @Test
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/requests/RequestsAPICreationTests.java:1630]: Can we just add an item in status `Available` to the next test, and remove this one?

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 4]
```diff
@@ -1625,6 +1627,104 @@ private void recallItem(ItemResource item, List<UserResource> users) {
...
+    Response response = requestsFixture.attemptPlaceHoldOrRecallTLR(firstItem.getInstanceId(),
+      usersFixture.charlotte(), RECALL);
+
+    assertThat(response.getStatusCode(), CoreMatchers.is(422));
+    assertThat(response.getJson(), hasErrorWith(
+      hasMessage("Request has no loan or recallable item")));
+  }
+
+  @Test
+  void tlrRecallShouldFailIfNoLoanAndInstanceHasItemsStatusAllowedForHold() {
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/requests/RequestsAPICreationTests.java:1644]:
```suggestion
  void tlrRecallShouldFailWhenRequestHasNoLoanOrRecallableItem() {
```

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 5]
```diff
@@ -1625,6 +1627,104 @@ private void recallItem(ItemResource item, List<UserResource> users) {
...
+    Response response = requestsFixture.attemptPlaceHoldOrRecallTLR(instanceId,
+      usersFixture.charlotte(), RECALL);
+
+    assertThat(response.getStatusCode(), CoreMatchers.is(422));
+    assertThat(response.getJson(), hasErrorWith(
+      hasMessage("Request has no loan or recallable item")));
+  }
+
+  @Test
+  void tlrRecallShouldPickOneFromAllowedStatusItemsIfLoanNotExistsAndInstanceHasAllowedItems() {
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/requests/RequestsAPICreationTests.java:1665]:
```suggestion
  void tlrRecallWithoutLoanShouldPickRecallableItemFromRequestedInstance() {
```

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 6]
```diff
@@ -1625,6 +1627,104 @@ private void recallItem(ItemResource item, List<UserResource> users) {
...
+      .withMaterialType(materialTypesFixture.book().getId())
+      .withPermanentLocation(locationsFixture.mainFloor())
+      .create());
+    requestsClient.create(buildPageTitleLevelRequest(usersFixture.jessica().getId(),
+      servicePointsFixture.cd1().getId(), instanceId));
+    itemsFixture.basedUponDunkirk(holdingBuilder -> holdingBuilder,
+      instanceBuilder -> instanceBuilder.withId(instanceId), ItemBuilder::awaitingDelivery);
+    itemsFixture.basedUponDunkirk(holdingBuilder -> holdingBuilder,
+      instanceBuilder -> instanceBuilder.withId(instanceId), ItemBuilder::awaitingPickup);
+
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/requests/RequestsAPICreationTests.java:1683]: redundant blank line

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 7]
```diff
@@ -3680,6 +3796,33 @@ public static IndividualResource setupMissingItem(ItemsFixture itemsFixture) {
...
+          is(ItemStatus.IN_PROCESS.getValue()));
+        break;
+      case MISSING:
+        item = itemsFixture.basedUponSmallAngryPlanet(ItemBuilder::missing);
+        assertThat(item.getResponse().getJson().getJsonObject("status").getString("name"),
+          is(ItemStatus.MISSING.getValue()));
+        break;
+    }
+    return item;
+  }
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/requests/RequestsAPICreationTests.java:3824]:
```suggestion
  private ItemResource createItemWithStatus(String itemStatus) {
    return itemsFixture.basedUponSmallAngryPlanet(
      builder -> builder.withStatus(itemStatus.getValue()));
  }
```

For this, you need to make method `api.support.builders.ItemBuilder#withStatus` public. No need to pass item fixture, it is an instance field on this test class.
You only use this method once. Can we just inline it?

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 8]
```diff
@@ -1625,6 +1627,104 @@ private void recallItem(ItemResource item, List<UserResource> users) {
...
+      .getString("status"))));
+    assertThat(response.getJson().getString("requestType"), is("Recall"));
+    assertThat(response.getJson().getString("requestLevel"), is("Title"));
+    assertThat(response.getJson().getString("instanceId"), is(instanceId));
+    assertThat(response.getJson().getString("fulfilmentPreference"), is("Hold Shelf"));
+    assertThat(response.getJson().getString("status"), is("Open - Not yet filled"));
+  }
+
+  @Test
+  void tlrRecallShouldPickItemWithAllowedStatusIfLoanNotExistsAndInstanceHasOtherItems() {
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/requests/RequestsAPICreationTests.java:1703]: Can we replace all your recall tests with one test where we:
- create items with every possible item status
- create recall requests until all items with recallable statuses are recalled (and verify every created request)
- verify that recall request fails when we are out of recallable items

File: src/test/java/api/requests/RequestsAPICreationTests.java (hunk 3 above)
Comment (bisecomsergiy) [reply src/test/java/api/requests/RequestsAPICreationTests.java]: Tests are refactored as per agreement, thanks.

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 9]
```diff
@@ -1625,6 +1627,104 @@ private void recallItem(ItemResource item, List<UserResource> users) {
       .forEach(index -> requestsFixture.recallItem(item, users.get(index)));
   }
 
+  @Test
+  void tlrRecallShouldFailIfLoanAndAllowedStatusItemsForInstanceNotExist() {
```
[the remark this answers is not in the document]
Comment (bisecomsergiy) [reply src/test/java/api/requests/RequestsAPICreationTests.java]: Tests are refactored as per agreement, thanks.

File: src/test/java/api/requests/RequestsAPICreationTests.java (hunk 2 above)
Comment (bisecomsergiy) [reply src/test/java/api/requests/RequestsAPICreationTests.java]: Removed, thanks.

File: src/test/java/api/requests/RequestsAPICreationTests.java (hunk 4 above)
Comment (bisecomsergiy) [reply src/test/java/api/requests/RequestsAPICreationTests.java]: Tests are refactored as per agreement, thanks.

File: src/test/java/api/requests/RequestsAPICreationTests.java (hunk 5 above)
Comment (bisecomsergiy) [reply src/test/java/api/requests/RequestsAPICreationTests.java]: Tests are refactored as per agreement, thanks.

File: src/test/java/api/requests/RequestsAPICreationTests.java (hunk 6 above)
Comment (bisecomsergiy) [reply src/test/java/api/requests/RequestsAPICreationTests.java]: Removed line, thanks.

File: src/test/java/api/requests/RequestsAPICreationTests.java (hunk 8 above)
Comment (bisecomsergiy) [reply src/test/java/api/requests/RequestsAPICreationTests.java]: Tests are refactored as per agreement, thanks.

File: src/test/java/api/requests/RequestsAPICreationTests.java (hunk 7 above)
Comment (bisecomsergiy) [reply src/test/java/api/requests/RequestsAPICreationTests.java]: It is inlined, thank you.

File: src/main/java/org/folio/circulation/resources/RequestFromRepresentationService.java [hunk 10]
```diff
@@ -272,16 +280,42 @@ private CompletableFuture<Result<Request>> fetchItemAndLoanForRecallTlrRequest(R
         requestQueue.getRecalledLoansIds())
```
Comment (Mykyta-Varenyk) [inline src/main/java/org/folio/circulation/resources/RequestFromRepresentationService.java:280]: Extract this as a variable and pass into shouldCheckRecalledLoan

File: src/test/java/api/support/builders/ItemBuilder.java [hunk 11]
```diff
@@ -157,11 +157,27 @@ public ItemBuilder withdrawn() {
...
+
+  public ItemBuilder inProcessNonRequestable() { return withStatus("In process (non-requestable)"); }
+
+  public ItemBuilder unavailable() { return withStatus("Unavailable"); }
+
+  public ItemBuilder restricted() { return withStatus("Restricted"); }
+
+  public ItemBuilder unknown() { return withStatus("Unknown"); }
+
+  public ItemBuilder orderClosed() { return withStatus("Order closed"); }
```
[the remark this answers is not in the document]
Comment (bisecomsergiy) [reply src/test/java/api/support/builders/ItemBuilder.java]: The methods are moved and used withStatus("Status") property, thanks.

File: src/main/java/org/folio/circulation/resources/RequestFromRepresentationService.java [hunk 12]
```diff
@@ -272,16 +280,42 @@ private CompletableFuture<Result<Request>> fetchItemAndLoanForRecallTlrRequest(R
...
       .thenApply(resultLoan -> resultLoan.map(request::withLoan))
-      .thenComposeAsync(requestResult -> requestResult.combineAfter(
-        r -> itemRepository.fetchFor(r.getLoan()), Request::withItem))
+      .thenCompose(r -> r.after(this::findItemForRecall))
       .thenComposeAsync(requestResult -> requestResult.combineAfter(
         this::getUserForExistingLoan, this::addUserToLoan))
       .thenApply(r -> errorHandler.handleValidationResult(r, INSTANCE_DOES_NOT_EXIST, request));
   }
 
+  private CompletableFuture<Result<Boolean>> shouldCheckRecalledLoan(Loan loan, RequestQueue requestQueue) {
```
[the remark this answers is not in the document]
Comment (bisecomsergiy) [reply src/main/java/org/folio/circulation/resources/RequestFromRepresentationService.java]: Name is changed, thanks

File: src/main/java/org/folio/circulation/resources/RequestFromRepresentationService.java (hunk 10 above)
Comment (bisecomsergiy) [reply src/main/java/org/folio/circulation/resources/RequestFromRepresentationService.java]: It is extracted, thanks.

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 13]
```diff
@@ -1828,11 +1903,30 @@ void canCreateHoldRequestWhenItemIsMissing() {
...
+  void canCreateTlrHoldRequestWhenInstanceHasItemsWithStatusAllowedForHold(String itemStatus) {
+    configurationsFixture.enableTlrFeature();
+    final ItemResource item = itemsFixture.basedUponSmallAngryPlanet(
+      builder -> builder.withStatus(itemStatus));
+    IndividualResource response = requestsFixture.placeTitleLevelHoldShelfRequest(item.getInstanceId(),
+      usersFixture.charlotte());
+
+    assertThat(response.getJson().getString("requestType"), is(HOLD.getValue()));
+    assertThat(response.getJson().getString("requestLevel"), is("Title"));
+    assertThat(response.getJson().getString("status"), is("Open - Not yet filled"));
```
Comment (Mykyta-Varenyk) [inline src/test/java/api/requests/RequestsAPICreationTests.java:1925]: Why is this assert made twice?

File: src/test/java/api/requests/RequestsAPICreationTests.java [hunk 14]
```diff
@@ -1625,6 +1625,81 @@ private void recallItem(ItemResource item, List<UserResource> users) {
...
+    Response response = requestsClient.attemptCreate(
+      new RequestBuilder()
+        .recall()
+        .withPickupServicePointId(requestPickupServicePoint.getId())
+        .titleRequestLevel()
+        .withInstanceId(instanceId)
+        .withNoItemId()
+        .withNoHoldingsRecordId()
+        .by(usersFixture.jessica())
+        .create());
```
Comment (Mykyta-Varenyk) [inline src/test/java/api/requests/RequestsAPICreationTests.java:1662]: please use placeTitleLevelRecallRequest

**Review by Mykyta-Varenyk — APPROVED** [review]

File: src/test/java/api/requests/RequestsAPICreationTests.java (hunk 13 above)
Comment (bisecomsergiy) [reply src/test/java/api/requests/RequestsAPICreationTests.java]: Typo, corrected, thanks.

File: src/test/java/api/requests/RequestsAPICreationTests.java (hunk 14 above)
Comment (bisecomsergiy) [reply src/test/java/api/requests/RequestsAPICreationTests.java]: Used placeTitleLevelRecallRequest, thanks

**Review by alexanderkurash — APPROVED** [review]

**Pull request merged**